src/mixins/java/org/spongepowered/common/mixin/core/entity/player/ServerPlayerEntityMixin.java

```diff
@@ -468,26 +495,144 @@ public void teleportTo(net.minecraft.world.server.ServerWorld world, double x, d
+                        }
+
+                        serverworld.getProfiler().pop();
+                        serverworld.getProfiler().push("placing");
+                        this.shadow$setLevel(targetWorld);
+                        targetWorld.addDuringPortalTeleport((ServerPlayerEntity) (Object) this);
+                        this.shadow$setRot(portalinfo.yRot, portalinfo.xRot);
+                        this.shadow$moveTo(portalinfo.pos.x, portalinfo.pos.y, portalinfo.pos.z);
+                        serverworld.getProfiler().pop();
+                        // Sponge Start: return this entity
+                        return (Entity) (Object) this;
+                    });
+                    // Sponge End
+                    // This is done here (it's not in Forge) to ensure that it's done. This is to ensure
+                    // advancements are triggered.
+                    this.shadow$triggerDimensionChangeTriggers(serverworld);
+                    // Sponge Start: make sure the right object was returned
+                    if (e != (Object) this) {
+                        throw new IllegalArgumentException(String.format("Teleporter %s "
+                                + "did not return the expected player entity: got %s, expected PlayerEntity %s", spongeTeleporter, e, this));
+                    }
+
+                    // If we need to reposition: well... reposition.
+                    // Downside: portals won't come with us, but with how it's implemented in Forge,
+                    // not sure how we'd do this
+                    final Vector3d destination = VecHelper.toVector3d(this.shadow$position());
+                    final ChangeEntityWorldEvent.Reposition reposition = SpongeEventFactory.createChangeEntityWorldEventReposition(
+                            PhaseTracker.getCauseStackManager().getCurrentCause(),
+                            (org.spongepowered.api.entity.Entity) this,
+                            (org.spongepowered.api.world.server.ServerWorld) serverworld,
+                            currentPosition,
+                            destination,
+                            (org.spongepowered.api.world.server.ServerWorld) originalDestinationWorld,
+                            originalDestination,
+                            (org.spongepowered.api.world.server.ServerWorld) targetWorld
+                    );
+                    if (reposition.isCancelled()) {
+                        // send them back to the original destination
+                        this.shadow$setPos(originalDestination.getX(), originalDestination.getY(), originalDestination.getZ());
+                    } else if (reposition.getDestinationPosition() != destination) {
+                        final Vector3d finalPosition = reposition.getDestinationPosition();
+                        this.shadow$setPos(finalPosition.getX(), finalPosition.getY(), finalPosition.getZ());
```

Comment:
So if we keep the strategy of doing reposition like this, we need to rollback the portal blocks in the PhaseTracker, issue a ticket for the chunk pos the player is going to, and call moveTo instead.